x-pack/plugins/timelines/server/search_strategy/timeline/factory/helpers/format_timeline_data.ts

```diff
@@ -5,118 +5,116 @@
-        value: '',
-        tiebreaker: null,
-      },
-    })
-  );
-
-const getValuesFromFields = async (
+const createBaseTimelineEdges = (): TimelineEdges => ({
+  node: {
+    ecs: { _id: '' },
+    data: [],
+    _id: '',
+    _index: '',
+  },
+  cursor: {
+    value: '',
+    tiebreaker: null,
+  },
+});
+
+const processMetadataField = (fieldName: string, hit: EventHit): TimelineNonEcsData[] => [
+  {
+    field: fieldName,
+    value: toStringArray(get(fieldName, hit)),
+  },
+];
+
+const processFieldData = (
   fieldName: string,
   hit: EventHit,
   nestedParentFieldName?: string
-): Promise<TimelineNonEcsData[]> => {
-  if (ECS_METADATA_FIELDS.includes(fieldName)) {
-    return [{ field: fieldName, value: toStringArray(get(fieldName, hit)) }];
-  }
+): TimelineNonEcsData[] => {
+  const fieldToEval = nestedParentFieldName
+    ? { [nestedParentFieldName]: hit.fields[nestedParentFieldName] }
+    : { [fieldName]: hit.fields[fieldName] };
 
-  let fieldToEval;
+  const formattedData = getDataFromFieldsHits(fieldToEval);
 
-  if (nestedParentFieldName == null) {
-    fieldToEval = {
-      [fieldName]: hit.fields[fieldName],
-    };
-  } else {
-    fieldToEval = {
-      [nestedParentFieldName]: hit.fields[nestedParentFieldName],
-    };
-  }
-  const formattedData = await getDataSafety(getDataFromFieldsHits, fieldToEval);
-  return formattedData.reduce((acc: TimelineNonEcsData[], { field, values }) => {
-    // nested fields return all field values, pick only the one we asked for
-    if (field.includes(fieldName)) {
-      acc.push({ field, value: values });
-    }
-    return acc;
-  }, []);
+  // Filter and map in one pass
+  return formattedData
+    .filter((item) => item.field.includes(fieldName))
+    .map(({ field, values }) => ({
+      field,
+      value: values,
+    }));
 };
 
-const mergeTimelineFieldsWithHit = async <T>(
-  fieldName: string,
-  flattenedFields: T,
-  hit: EventHit,
-  dataFields: readonly string[],
-  ecsFields: readonly string[]
-) => {
-  if (fieldName != null) {
-    const nestedParentPath = getNestedParentPath(fieldName, hit.fields);
-    if (
-      nestedParentPath != null ||
-      has(fieldName, hit.fields) ||
-      ECS_METADATA_FIELDS.includes(fieldName)
-    ) {
-      const objectWithProperty = {
-        node: {
-          ...get('node', flattenedFields),
-          data: dataFields.includes(fieldName)
-            ? [
-                ...get('node.data', flattenedFields),
-                ...(await getValuesFromFields(fieldName, hit, nestedParentPath)),
-              ]
-            : get('node.data', flattenedFields),
-          ecs: ecsFields.includes(fieldName)
-            ? {
-                ...get('node.ecs', flattenedFields),
-                ...buildObjectRecursive(fieldName, hit.fields),
-              }
-            : get('node.ecs', flattenedFields),
-        },
-      };
-      return merge(flattenedFields, objectWithProperty);
+export const formatTimelineData = async (
```

Comment:
does it have to be `async`? :) it could be though, if the amount of data to process justifies splitting things into chunks

Reply:
ya it should be, if kibana is under load and the number of fields is very large. I think that was the case the code worked like this in the first place, however it's happening too much, with too many nested reduces/spreads/etc, slowing the response down even with a moderate number of fields.